File: macos/BlindfoldCore/Tests/BlindfoldCoreTests/AppStateMachineTests.swift
import Testing
@testable import BlindfoldCore

/// The five-state machine (ADR-0039): fused from proxy liveness (owned by the
/// supervisor) and `/v1/status`'s `state`. Pure reduction — feed status payloads +
/// liveness signals, assert the resulting state. No UI, no I/O.
///
/// Cases are the shared golden-vector fixture (issue #193 / ADR-0041) — the same
/// reducer truth table the future C# `Blindfold.Core` (#194) asserts against, so
/// the two cores can't silently drift.
@Test(arguments: GoldenVectorFixture.load().reducer_truth_table)
func reducerMatchesGoldenVector(_ vector: GoldenVectorFixture.ReducerCase) {
    let state = AppStateMachine.reduce(
        liveness: vector.liveness.toLiveness(),
        status: vector.status?.toStatusPayload()
    )
    #expect(state == vector.expected_state.toAppState(), "\(vector.name)")
}

/// The Unprotected alarm (ADR-0038) overlays the five-state machine rather than
/// replacing it — read verbatim from #180's `unprotected_mode` status fields.
@Test(arguments: GoldenVectorFixture.load().unprotected_alarm_cases)
func unprotectedAlarmMatchesGoldenVector(_ vector: GoldenVectorFixture.UnprotectedAlarmCase) {
    let alarm = AppStateMachine.unprotectedAlarm(status: vector.status.toStatusPayload())
    #expect(alarm == vector.expected_alarm?.toAlarm(), "\(vector.name)")
}

File: windows/Blindfold.Core.Tests/AppStateMachineTests.cs
using Blindfold.Core;
using Xunit;

namespace Blindfold.Core.Tests;

/// <summary>
/// The menu-bar/tray five-state machine (ADR-0039/0041): fused from <c>ProxyLiveness</c>
/// (the supervisor) and <c>/v1/status</c>'s <c>state</c>. Cases are the shared
/// golden-vector fixture (issue #193 / ADR-0041) — the same reducer truth table the Swift
/// <c>AppStateMachine</c> is held to.
/// </summary>
public class AppStateMachineTests
{
    public static IEnumerable<object[]> ReducerCases() =>
        GoldenVectorFixture.Load().ReducerTruthTable.Select(c => new object[] { c });

    [Theory]
    [MemberData(nameof(ReducerCases))]
    public void ReducerMatchesGoldenVector(GoldenVectorFixture.ReducerCase vector)
    {
        var liveness = vector.Liveness.ToLiveness();
        var status = vector.Status?.ToStatusPayload();
        var expected = vector.ExpectedState.ToAppState();

        var actual = AppStateMachine.Reduce(liveness, status);

        Assert.Equal(expected, actual);
    }

    public static IEnumerable<object[]> UnprotectedAlarmCases() =>
        GoldenVectorFixture.Load().UnprotectedAlarmCases.Select(c => new object[] { c });

    [Theory]
    [MemberData(nameof(UnprotectedAlarmCases))]
    public void UnprotectedAlarmMatchesGoldenVector(GoldenVectorFixture.UnprotectedAlarmCase vector)
    {
        var status = vector.Status.ToStatusPayload();
        var expected = vector.ExpectedAlarm?.ToAlarm();

        var actual = AppStateMachine.UnprotectedAlarmFor(status);

        Assert.Equal(expected, actual);
    }
}
